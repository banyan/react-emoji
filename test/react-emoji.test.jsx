import React, { PropTypes } from 'react'
import { mount } from 'enzyme'

import ReactEmoji, { emojify } from '../src/index'

const SampleComponent = ({ text, options }) => (
  <div>
    <span>{ emojify(text, options) }</span>
  </div>
)

SampleComponent.propTypes = {
  text: PropTypes.string.isRequired,
  options: PropTypes.shape({
    useEmoticon: PropTypes.bool,
    emojiType: PropTypes.string,
    host: PropTypes.string,
    path: PropTypes.string,
    ext: PropTypes.string,
    attributes: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    singleEmoji: PropTypes.bool,
    strict: PropTypes.bool,
  })
}

SampleComponent.defaultProps = {
  options: {}
}

describe('ReactEmoji', () => {
  it('renders annotation', () => {
    const wrapper = mount(
      <ReactEmoji>:smile:</ReactEmoji>
    )
    expect(wrapper).toMatchSnapshot('renders :smile: annotation with ReactEmoji')
  })
})

describe('emojify', () => {
  context('parse', () => {
    it('renders annotation', () => {
      const wrapper = mount(
        <SampleComponent
          text=":smile:"
        />,
      )
      expect(wrapper).toMatchSnapshot('renders :smile: annotation')
    })

    it('renders emoticon', () => {
      const wrapper = mount(
        <SampleComponent
          text=":("
        />,
      )
      expect(wrapper).toMatchSnapshot('renders :( emoticon')
    })

    context(':/', () => { // https://github.com/banyan/react-emoji/issues/1
      it('converts to emoji', () => {
        const wrapper = mount(
          <SampleComponent
            text=":/"
          />,
        )
        expect(wrapper).toMatchSnapshot('renders :/ emoticon')
      })

      it('does not convert to emoji', () => {
        const wrapper = mount(
          <SampleComponent
            text="http://example.org"
          />,
        )
        expect(wrapper).toMatchSnapshot('does not render :/ emoticon')
      })
    })

    context('when emoji key is only for GitHub', () => {
      it('does not convert to emoji', () => {
        const wrapper = mount(
          <SampleComponent
            text=":octocat:"
          />,
        )
        expect(wrapper).toMatchSnapshot('does not render :octocat: annotation')
      })
    })
  })

  context('options', () => {
    context('useEmoticon', () => {
      it('renders emoticon', () => {
        const wrapper = mount(
          <SampleComponent
            text=":)"
            options={{ useEmoticon: true }}
          />,
        )
        expect(wrapper).toMatchSnapshot('renders :) emoticon')
      })

      it('does not render emoticon', () => {
        const wrapper = mount(
          <SampleComponent
            text=":)"
            options={{ useEmoticon: false }}
          />,
        )
        expect(wrapper).toMatchSnapshot('does not render :) emoticon')
      })
    })

    context('emojiType', () => {
      it("uses emojione's emoji", () => {
        const wrapper = mount(
          <SampleComponent
            text=":)"
            options={{ emojiType: 'emojione' }}
          />,
        )
        expect(wrapper).toMatchSnapshot("uses emojione's emoji")
      })
    })

    context('host', () => {
      it('uses given host', () => {
        const wrapper = mount(
          <SampleComponent
            text=":)"
            options={{ host: 'http://example.org' }}
          />,
        )
        expect(wrapper).toMatchSnapshot('uses given host')
      })
    })

    context('path', () => {
      it('uses given path', () => {
        const wrapper = mount(
          <SampleComponent
            text=":)"
            options={{ host: 'http://example.org', path: 'foo/bar' }}
          />,
        )
        expect(wrapper).toMatchSnapshot('uses given host and path')
      })
    })

    context('ext', () => {
      it('uses given ext', () => {
        const wrapper = mount(
          <SampleComponent
            text=":)"
            options={{ ext: 'png' }}
          />,
        )
        expect(wrapper).toMatchSnapshot('uses given ext')
      })
    })

    context('attributes', () => {
      it('uses given attributes', () => {
        const wrapper = mount(
          <SampleComponent
            text=":)"
            options={{ attributes: { width: '30px', className: 'foo' } }}
          />,
        )
        expect(wrapper).toMatchSnapshot('uses given attributes')
      })
    })

    context('singleEmoji', () => {
      it('renders annotation', () => {
        const wrapper = mount(
          <SampleComponent
            text=":smile:"
            options={{ singleEmoji: true }}
          />,
        )
        expect(wrapper).toMatchSnapshot('renders :smile: annotation')
      })

      it('renders emoticon', () => {
        const wrapper = mount(
          <SampleComponent
            text=":)"
            options={{ singleEmoji: true }}
          />,
        )
        expect(wrapper).toMatchSnapshot('renders :) emoticon')
      })

      it('shows text as given if emoji is not found', () => {
        const wrapper = mount(
          <SampleComponent
            text=":foobarbaz:"
            options={{ singleEmoji: true }}
          />,
        )
        expect(wrapper).toMatchSnapshot('renders :foobarbaz: as text')
      })
    })

    context('strict', () => {
      context('strict is true', () => {
        context('if emoji is found on list', () => {
          it('renders emoji', () => {
            const wrapper = mount(
              <SampleComponent
                text=":smile:"
                options={{ strict: true }}
              />,
            )
            expect(wrapper).toMatchSnapshot('renders :smile: annotation')
          })
        })

        context('if emoji is not found on list', () => {
          context('singleEmoji is not given', () => {
            it('throws an error', () => {
              const wrapper = () => (
                mount(
                  <SampleComponent
                    text=":foobarbaz:"
                    options={{ strict: true }}
                  />,
                )
              )
              expect(wrapper).toThrow(new Error('Could not find emoji: :foobarbaz:.'))
            })
          })

          context('singleEmoji is given', () => {
            it('throws an error', () => {
              const wrapper = () => (
                mount(
                  <SampleComponent
                    text=":foobarbaz:"
                    options={{ strict: true, singleEmoji: true }}
                  />,
                )
              )
              expect(wrapper).toThrow(new Error('Could not find emoji: :foobarbaz:.'))
            })
          })
        })
      })
    })
  })

  context('multiple text', () => {
    it('renders emoticon', () => {
      const wrapper = mount(
        <SampleComponent
          text="foo :smile: :("
        />,
      )
      expect(wrapper).toMatchSnapshot('renders multiple emoticon and annotation')
    })
  })

  context('custom emoji', () => {
    it('renders from custom dictionary', () => {
      const customDict = {
        batman: "https://my.custom.server.com/batman.jpg"
      };
      const wrapper = mount(
        <SampleComponent
          text="foo :batman: bar"
          options={{ customDict: customDict }}
        />,
      )
      expect(wrapper).toMatchSnapshot('renders custom emoji')
    })
  })
})
