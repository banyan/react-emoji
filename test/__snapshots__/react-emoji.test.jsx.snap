// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render :) emoticon 1`] = `
<SampleComponent
  options={
    Object {
      "useEmoticon": false,
    }
  }
  text=":)"
>
  <div>
    <span>
      :)
    </span>
  </div>
</SampleComponent>
`;

exports[`does not render :/ emoticon 1`] = `
<SampleComponent
  options={Object {}}
  text="http://example.org"
>
  <div>
    <span>
      http://example.org
    </span>
  </div>
</SampleComponent>
`;

exports[`does not render :octocat: annotation 1`] = `
<SampleComponent
  options={Object {}}
  text=":octocat:"
>
  <div>
    <span>
      :octocat:
    </span>
  </div>
</SampleComponent>
`;

exports[`renders :( emoticon 1`] = `
<SampleComponent
  options={Object {}}
  text=":("
>
  <div>
    <span>
      <img
        height="20px"
        src="https://twemoji.maxcdn.com/svg/1f61e.svg"
        width="20px"
      />
    </span>
  </div>
</SampleComponent>
`;

exports[`renders :) emoticon 1`] = `
<SampleComponent
  options={
    Object {
      "useEmoticon": true,
    }
  }
  text=":)"
>
  <div>
    <span>
      <img
        height="20px"
        src="https://twemoji.maxcdn.com/svg/1f642.svg"
        width="20px"
      />
    </span>
  </div>
</SampleComponent>
`;

exports[`renders :) emoticon 2`] = `
<SampleComponent
  options={
    Object {
      "singleEmoji": true,
    }
  }
  text=":)"
>
  <div>
    <span>
      <img
        height="20px"
        src="https://twemoji.maxcdn.com/svg/1f642.svg"
        width="20px"
      />
    </span>
  </div>
</SampleComponent>
`;

exports[`renders :/ emoticon 1`] = `
<SampleComponent
  options={Object {}}
  text=":/"
>
  <div>
    <span>
      <img
        height="20px"
        src="https://twemoji.maxcdn.com/svg/1f615.svg"
        width="20px"
      />
    </span>
  </div>
</SampleComponent>
`;

exports[`renders :foobarbaz: as text 1`] = `
<SampleComponent
  options={
    Object {
      "singleEmoji": true,
    }
  }
  text=":foobarbaz:"
>
  <div>
    <span>
      :foobarbaz:
    </span>
  </div>
</SampleComponent>
`;

exports[`renders :smile: annotation 1`] = `
<SampleComponent
  options={Object {}}
  text=":smile:"
>
  <div>
    <span>
      <img
        height="20px"
        src="https://twemoji.maxcdn.com/svg/1f604.svg"
        width="20px"
      />
    </span>
  </div>
</SampleComponent>
`;

exports[`renders :smile: annotation 2`] = `
<SampleComponent
  options={
    Object {
      "singleEmoji": true,
    }
  }
  text=":smile:"
>
  <div>
    <span>
      <img
        height="20px"
        src="https://twemoji.maxcdn.com/svg/1f604.svg"
        width="20px"
      />
    </span>
  </div>
</SampleComponent>
`;

exports[`renders :smile: annotation 3`] = `
<SampleComponent
  options={
    Object {
      "strict": true,
    }
  }
  text=":smile:"
>
  <div>
    <span>
      <img
        height="20px"
        src="https://twemoji.maxcdn.com/svg/1f604.svg"
        width="20px"
      />
    </span>
  </div>
</SampleComponent>
`;

exports[`renders :smile: annotation with ReactEmoji 1`] = `
<ReactEmoji>
  <img
    height="20px"
    src="https://twemoji.maxcdn.com/svg/1f604.svg"
    width="20px"
  />
</ReactEmoji>
`;

exports[`renders custom emoji 1`] = `
<SampleComponent
  options={
    Object {
      "customDict": Object {
        "batman": "https://my.custom.server.com/batman.jpg",
      },
    }
  }
  text="foo :batman: bar"
>
  <div>
    <span>
      foo 
      <img
        height="20px"
        src="https://my.custom.server.com/batman.jpg"
        width="20px"
      />
       bar
    </span>
  </div>
</SampleComponent>
`;

exports[`renders multiple emoticon and annotation 1`] = `
<SampleComponent
  options={Object {}}
  text="foo :smile: :("
>
  <div>
    <span>
      foo 
      <img
        height="20px"
        src="https://twemoji.maxcdn.com/svg/1f604.svg"
        width="20px"
      />
       
      <img
        height="20px"
        src="https://twemoji.maxcdn.com/svg/1f61e.svg"
        width="20px"
      />
    </span>
  </div>
</SampleComponent>
`;

exports[`uses emojione's emoji 1`] = `
<SampleComponent
  options={
    Object {
      "emojiType": "emojione",
    }
  }
  text=":)"
>
  <div>
    <span>
      <img
        height="20px"
        src="https://cdn.jsdelivr.net/emojione/assets/svg/1F642.svg"
        width="20px"
      />
    </span>
  </div>
</SampleComponent>
`;

exports[`uses given attributes 1`] = `
<SampleComponent
  options={
    Object {
      "attributes": Object {
        "className": "foo",
        "width": "30px",
      },
    }
  }
  text=":)"
>
  <div>
    <span>
      <img
        className="foo"
        height="20px"
        src="https://twemoji.maxcdn.com/svg/1f642.svg"
        width="30px"
      />
    </span>
  </div>
</SampleComponent>
`;

exports[`uses given ext 1`] = `
<SampleComponent
  options={
    Object {
      "ext": "png",
    }
  }
  text=":)"
>
  <div>
    <span>
      <img
        height="20px"
        src="https://twemoji.maxcdn.com/png/1f642.png"
        width="20px"
      />
    </span>
  </div>
</SampleComponent>
`;

exports[`uses given host 1`] = `
<SampleComponent
  options={
    Object {
      "host": "http://example.org",
    }
  }
  text=":)"
>
  <div>
    <span>
      <img
        height="20px"
        src="http://example.org/1f642.svg"
        width="20px"
      />
    </span>
  </div>
</SampleComponent>
`;

exports[`uses given host and path 1`] = `
<SampleComponent
  options={
    Object {
      "host": "http://example.org",
      "path": "foo/bar",
    }
  }
  text=":)"
>
  <div>
    <span>
      <img
        height="20px"
        src="http://example.org/foo/bar/1f642.svg"
        width="20px"
      />
    </span>
  </div>
</SampleComponent>
`;
